pubsub/pubsub_test.go

```diff
@@ -0,0 +1,379 @@
+}
+
+type testResponseMarshaller struct{}
+
+func (t *testResponseMarshaller) Marshal(val string) []byte {
+	return []byte(val)
+}
+
+func (t *testResponseMarshaller) Unmarshal(val []byte) (string, error) {
+	return string(val), nil
+}
+
+func createGroup(ctx context.Context, t *testing.T, streamName, groupName string, client redis.UniversalClient) {
+	t.Helper()
+	_, err := client.XGroupCreateMkStream(ctx, streamName, groupName, "$").Result()
+	if err != nil {
+		t.Fatalf("Error creating stream group: %v", err)
+	}
+}
+
+type configOpt interface {
+	apply(consCfg *ConsumerConfig, prodCfg *ProducerConfig)
+}
+
+type disableReproduce struct{}
+
+func (e *disableReproduce) apply(_ *ConsumerConfig, prodCfg *ProducerConfig) {
+	prodCfg.EnableReproduce = false
+}
+
+func producerCfg() *ProducerConfig {
+	return &ProducerConfig{
+		EnableReproduce:      DefaultTestProducerConfig.EnableReproduce,
+		CheckPendingInterval: DefaultTestProducerConfig.CheckPendingInterval,
+		KeepAliveTimeout:     DefaultTestProducerConfig.KeepAliveTimeout,
+		CheckResultInterval:  DefaultTestProducerConfig.CheckResultInterval,
+	}
+}
+
+func consumerCfg() *ConsumerConfig {
+	return &ConsumerConfig{
+		ResponseEntryTimeout: DefaultTestConsumerConfig.ResponseEntryTimeout,
+		KeepAliveTimeout:     DefaultTestConsumerConfig.KeepAliveTimeout,
+	}
+}
+
+func newProducerConsumers(ctx context.Context, t *testing.T, opts ...configOpt) (*Producer[string, string], []*Consumer[string, string]) {
+	t.Helper()
+	redisURL := redisutil.CreateTestRedis(ctx, t)
+	prodCfg, consCfg := producerCfg(), consumerCfg()
+	prodCfg.RedisURL, consCfg.RedisURL = redisURL, redisURL
+	streamName := uuid.NewString()
+	groupName := fmt.Sprintf("group_%s", streamName)
+	prodCfg.RedisGroup, consCfg.RedisGroup = groupName, groupName
+	prodCfg.RedisStream, consCfg.RedisStream = streamName, streamName
+	for _, o := range opts {
+		o.apply(consCfg, prodCfg)
+	}
+	producer, err := NewProducer[string, string](prodCfg, &testRequestMarshaller{}, &testResponseMarshaller{})
+	if err != nil {
+		t.Fatalf("Error creating new producer: %v", err)
+	}
+
+	var consumers []*Consumer[string, string]
+	for i := 0; i < consumersCount; i++ {
+		c, err := NewConsumer[string, string](ctx, consCfg, &testRequestMarshaller{}, &testResponseMarshaller{})
+		if err != nil {
+			t.Fatalf("Error creating new consumer: %v", err)
+		}
+		consumers = append(consumers, c)
+	}
+	createGroup(ctx, t, streamName, groupName, producer.client)
+	return producer, consumers
+}
+
+func messagesMaps(n int) []map[string]string {
+	ret := make([]map[string]string, n)
+	for i := 0; i < n; i++ {
+		ret[i] = make(map[string]string)
+	}
+	return ret
+}
+
+func wantMessages(n int) []string {
+	var ret []string
+	for i := 0; i < n; i++ {
+		ret = append(ret, fmt.Sprintf("msg: %d", i))
+	}
+	sort.Slice(ret, func(i, j int) bool {
+		return fmt.Sprintf("%v", ret[i]) < fmt.Sprintf("%v", ret[j])
+	})
+	return ret
+}
+
+func TestRedisProduce(t *testing.T) {
```

Comment:
This test didn't really do same logic as other (e.g. this produces sequentially, blocks for response, producing messages was started after consumers, small differences for completeness).

Changed to use those helpers, and merged with the other test since code becomes duplicate that way.